### ISSUE TITLE
Segmentation fault with SharedArrays when --threads is set

### ISSUE THREAD

**Issue by Socob** [JuliaLang/julia#54253]
I’m getting segmentation faults when using `SharedArrays` and passing `--threads` to Julia, even when I’m not actually using any of those threads (see the MWE below). Using the commented line instead (without `--threads`), I’m not getting any segmentation faults.

Triggering the segfault does seem to depend on the number of worker processes, in that with a small number of workers, the issue is not triggered (or at least not consistently). The details may be machine-specific, but I’ve reproduced this on several different machines.

I don’t have any attempts at an explanation, since I don’t see how merely setting the number of Julia threads would affect this code.

---

1. The output of `versioninfo()`:
   >     Julia Version 1.10.2
   >     Commit bd47eca2c8a (2024-03-01 10:14 UTC)
   >     Build Info:
   >       Official https://julialang.org/ release
   >     Platform Info:
   >       OS: Linux (x86_64-linux-gnu)
   >       CPU: 16 × AMD Ryzen 7 4800H with Radeon Graphics
   >       WORD_SIZE: 64
   >       LIBM: libopenlibm
   >       LLVM: libLLVM-15.0.7 (ORCJIT, znver2)
   >     Threads: 1 default, 0 interactive, 1 GC (on 16 virtual cores)
2. How you installed Julia: juliaup
3. A minimal working example (MWE), also known as a minimum reproducible example:
    ```julia
    using Distributed
    
    function main()
        sa = SharedArray{Float64}(1000, 10000)
        fill!(sa, 0)
        println("write"); flush(stdout)
        sa .= 1.0
        println("read"); flush(stdout)
        @everywhere workers() begin
            # dummy calculation which reads from sa a lot
            sa = $sa
            for i in 1:size(sa, 2)
                sum(
                    sum(1.1 .* @view sa[:, i])
                    for _ in 1:2000
                )
            end
        end
        println("DONE"); flush(stdout)
    end
    
    addprocs(
        14;
        topology=:all_to_all, lazy=false,
        # results in segfault
        exeflags=`--startup-file=no --threads=16`
        # no segfault!
    #    exeflags=`--startup-file=no`
    )
    @everywhere begin
        using Distributed
        using SharedArrays
    end
    main()
    ```

**Comment by Socob** [comment]
OK, it’s definitely happening even for a normal array, so this has nothing to do with `SharedArrays`! That makes it much worse!

**Comment by danspielman** [comment]
I am having a similar error. My code runs fine on a Mac, but gives me errors like this when I run it on a linux cluster. I’ll try to construct an MFE — minimal failing example.

**Comment by LHerviou** [comment]
I am also having similar issues on a linux cluster (code runs fine on my old julia install on the laptop).
I put some data in https://discourse.julialang.org/t/segmentation-fault-using-multithreaded-julia-on-new-server/114557
Not able to get a clean MFE.

**Comment by wgmitchener** [comment]
I've been seeing this. Fedora Linux 41, x86_64. Julia 1.10.9. Here's a backtrace from a worker process that crashed. Each worker seems to be crashing in the same place. The job involves using SymPy to simplify and evaluate symbolic expressions, which is implemented with PyCall. The crash seems to have something to do with garbage collection. It looks like a finalizer that calls into python to decrement a refcount is where the bad memory reference occurs. The worker task runs fine without --threads.

[5609-worker-5 stderr.txt](https://github.com/user-attachments/files/19218025/5609-worker-5.stderr.txt)

**Comment by wgmitchener** [comment]
Possibly, but it's also possible that we might both be seeing different aspects of a bug in garbage collection that shows up in worker processes with threads. These sporadic crashes could be due to a race condition somewhere.

I haven't been able to get your test script to generate a segfault. My workstation doesn't have enough CPUs to run it as posted, and when I reduce the number of workers and threads, it completes without error.

**Comment by Socob** [comment]
Yeah, that’s the difficult part about this… When I originally reported this, I had a hard time seeing if anything made it more or less likely to trigger (beyond just increasing the number of threads). What happens when you oversubscribe your CPU? Even with fewer cores, you can still use a large number of threads, since we’re not looking for performance here.

**Comment by wgmitchener** [comment]
I had to change the backticks in exeflags= to regular quotes to make it work, by the way. Even with oversubscribing, the script completes without error.

Here's my versioninfo():

```
julia> versioninfo()
Julia Version 1.10.9
Commit 5595d20a287 (2025-03-10 12:51 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 20 × Intel(R) Xeon(R) Silver 4210R CPU @ 2.40GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, cascadelake)
Threads: 1 default, 0 interactive, 1 GC (on 20 virtual cores)
```
